fetcher: add filter to remove old blocks if new overlapping blocks found

khyatisoneji opened the pull request:
## Changes

Add Filter to filter out old blocks if new overlapping blocks found that fully submatch the old blocks

Please let me know if this is the right direction to go in terms of filtering out the old blocks

bwplotka commented on pkg/block/fetcher.go:
```diff
@@ -410,3 +411,52 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
+	for id, meta := range metas {
+		sources := make([]ulid.ULID, len(metas))
+		for _, sourceID := range meta.Compaction.Sources {
+			if sourceID == id {
```
Let's use `sourceID.Compare(id) == 0` to be safe maybe (:

bwplotka commented on pkg/block/fetcher.go:
```diff
@@ -410,3 +411,52 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
+		if areSlicesEqual(sources, meta.Compaction.Sources) {
+			for _, sourceID := range meta.Compaction.Sources {
+				if sourceID != id {
+					delete(metas, sourceID)
```
Next to deleting, we probably want store the deleted metas and check if we are ready to delete those blocks from bucket (:

bwplotka commented on pkg/block/fetcher.go:
```diff
@@ -410,3 +411,52 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
 	}
 }
+
+// CompactedBlockFilter is a MetaFetcher filter that filters out blocks that have newer block
```
Not addressed.

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +401,271 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+					},
+				},
+			},
+			expected: func(input map[ulid.ULID]*metadata.Meta) map[ulid.ULID]*metadata.Meta {
```
Why this complex signature for expected field? Why not just expect certain slice of ulid.ULID? (:

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +401,271 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+		expected func(input map[ulid.ULID]*metadata.Meta) map[ulid.ULID]*metadata.Meta
+	}{
+		{
+			name: "3 non compacted metas in bucket",
```
It does not matter if they are really compacted .. It still does not say to me if it's overlapping or not

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +401,271 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+			},
+		},
+		{
+			name: "compacted meta in bucket",
```
meta is compacted? Let's be precise (:

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +401,271 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+		{
+			name: "two blocks with same sources",
+			input: map[ulid.ULID]*metadata.Meta{
+				ULID(3): {
```
Configuring test cases should be clear and simple. I think we don't need to have the whole `metadata.Meta` We can just configure as input: `map[ulid.ULID]tsdb.BlockMetaCompaction` and that's it. We don't really configure anything else. Then we can create a proper map to pass to filter later on. 

However, piece with test cases should be concise and without many repetitions.

bwplotka replied:
not addressed

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +401,271 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+			},
+		},
+		{
+			name: "3 non compacted metas and compacted meta of level 2 in bucket",
```
again, meta is never technically compacted -> we need to consistent and clear in naming here. Also compaction does not matter here: Just what constains same sources.

For example this case could be named: `Block 4 contains 1, 2, 3`

Also I believe :
```
ULID: ULID(4),
						Compaction: tsdb.BlockMetaCompaction{
							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
						},
```
.. is wrong. This will never happen as 4 is compacted block, so it will be never part of sources I believe (That's why I linked the Prometheus code which does it, to double check) (:

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +401,271 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+			},
+		},
+		{
+			name: "compacted blocks with overlapping sources",
```
How is this different to other cases?

bwplotka replied:
ditto

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +401,271 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+			},
+		},
+	} {
+		f := NewCompactedBlockFilter()
```
Can we add case with 2 blocks having exactly the same sources?

khyatisoneji replied:
Yes, its added with test case name: `two compacted blocks with same sources`

bwplotka commented on pkg/block/forest.go:
```diff
@@ -0,0 +1,121 @@
+		rootSources := metas[vertex.ID].Compaction.Sources
+		sources := metas[id].Compaction.Sources
+
+		if f.containsSources(rootSources, sources) && f.containsSources(sources, rootSources) { // a block exists with same sources.
```
This line is quite long so maybe we can put comment above? And make it a full sentence? (Capital first letter)

bwplotka commented on pkg/block/forest.go:
```diff
@@ -0,0 +1,121 @@
+	for _, a := range s2 {
+		found := false
+		for _, e := range s1 {
+			if a.Compare(e) == 0 {
```
Nice. I think we can optimize a bit but this willl do for now (: 

We can create some go benchmark for this in next PR :muscle:

khyatisoneji commented on cmd/thanos/compact.go:
```diff
@@ -247,9 +247,11 @@ func runCompact(
 	metaFetcher, err := block.NewMetaFetcher(logger, 32, bkt, "", extprom.WrapRegistererWithPrefix("thanos_", reg),
 		block.NewLabelShardedMetaFilter(relabelConfig).Filter,
 		(&consistencyDelayMetaFilter{logger: logger, consistencyDelay: consistencyDelay}).Filter,
+		compactedBlockFilter.Filter,
```
Syncer is not used by store component, so we need to add the filter there explicitly then

khyatisoneji commented on cmd/thanos/compact.go:
```diff
@@ -247,9 +247,11 @@ func runCompact(
 		}
 	}()
 
+	compactedBlockFilter := block.NewCompactedBlockFilter()
```
renamed this to duplicatefilter.

khyatisoneji commented on pkg/block/fetcher.go:
```diff
@@ -410,3 +411,49 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
+// CompactedBlockFilter is a MetaFetcher filter that filters out blocks that have newer block
+// from two or more overlapping blocks that fully submatches the source blocks of the older blocks.
+type CompactedBlockFilter struct {
+	IdsToDelete []ulid.ULID
```
renamed this as DuplicateIDs

khyatisoneji commented on pkg/block/fetcher.go:
```diff
@@ -410,3 +411,49 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
+		forest.AddMeta(meta.ULID, metas)
+	}
+
+	ulidsToDelete := forest.GetNonRootIDs()
```
It is returning slice, can you please give more context on your concern

khyatisoneji commented:
@bwplotka I have addressed some concerns that you have made:
1. Removing forest and having only vertex
2. Name test cases with compacted blocks and not compacted metas.
3. Removed the function for expected metas, currently only comparing the slices.
4. Renamed ulidsToDelete with duplicateIds
5. Rename CompactedBlockFilter with DeduplicateFilter

I have one suggestion that we should address deletion in a separate PR. Let's narrow down scope of this PR to only filter out the duplicate blocks.
Deletion of the blocks, as you mentioned should be done with checks mentioned in the proposal, so adding them in a separate PR would be better.

bwplotka commented:
> I have one suggestion that we should address deletion in a separate PR. Let's narrow down scope of this PR to only filter out the duplicate blocks.
Deletion of the blocks, as you mentioned should be done with checks mentioned in the proposal, so adding them in a separate PR would be better.

Very good idea. But in this case we should not probably hook this filter yet in Compactor I think. We can hook in Store Gateway though.

bwplotka commented on pkg/block/fetcher.go:
```diff
@@ -410,3 +411,48 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
+// Filter filters out duplicate blocks that can be formed
+// from two or more overlapping blocks that fully submatches the source blocks of the older blocks.
+func (f *DeduplicateFilter) Filter(metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) {
+	forest := NewVertex(ulid.MustNew(uint64(0), nil), nil)
```
still forest name, maybe just `root`?

bwplotka commented on pkg/block/forest.go:
```diff
@@ -0,0 +1,109 @@
+)
+
+// Vertex type implements a vertex of a Forest.
+type Vertex struct {
```
filename is still old (forest)

bwplotka commented on pkg/block/forest.go:
```diff
@@ -0,0 +1,109 @@
+	return ulids
+}
+
+func (f *Vertex) add(id ulid.ULID, startVertex *Vertex, metas map[ulid.ULID]*metadata.Meta) bool {
```
`add` is still cryptic. At least can we have a comment what this method does? (: Because it's not just vertex - it's a vertex that 

Also:
```suggestion
func (v *Vertex) add(id ulid.ULID, startVertex *Vertex, metas map[ulid.ULID]*metadata.Meta) bool {
```

bwplotka commented on pkg/block/forest.go:
```diff
@@ -0,0 +1,109 @@
+	return f.add(id, rootVertex, metas)
+}
+
+func (f *Vertex) containsSources(rootSources, sources []ulid.ULID) bool {
```
Does it have to be method? Also, I don't get why we need this at all this method does nothing ): It might too shallow

bwplotka commented on pkg/block/fetcher.go:
```diff
@@ -410,3 +411,48 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
 	}
 }
+
+// DeduplicateFilter is a MetaFetcher filter that filters out duplicate blocks.
```
```suggestion
// DeduplicateFilter is a MetaFetcher filter that filters out older blocks that have exactly the same data.
```

bwplotka commented on pkg/block/fetcher.go:
```diff
@@ -410,3 +411,48 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
+	})
+
+	for _, meta := range metaSlice {
+		root.AddMeta(meta.ULID, metas)
```
Still, I would vote to move it as a separate function (:

bwplotka commented on pkg/block/node.go:
```diff
@@ -0,0 +1,104 @@
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+)
+
+// Node type implements a node of a Forest.
```
forest still ):

khyatisoneji replied:
Suggestions? Node of a tree?

khyatisoneji replied:
Aren't both test cases relevant? The first one is for case where two blocks have same sources and second one is where the blocks have overlapping sources.

bwplotka commented on pkg/block/node.go:
```diff
@@ -0,0 +1,87 @@
+	return ulids
+}
+
+func (n *Node) add(id ulid.ULID, startNode *Node, metas map[ulid.ULID]*metadata.Meta, areEqual, isChildNode compareNodes) bool {
```
Ok, I said like 4 times to move this from node and it's still there and you did not explain why you are against this (: 

Maybe I can send PR to your PR to show you how it can be done and show adventages? (:

bwplotka replied:
The main problem is that this is just a `node` 

It's not  `hierarchicalBlocksNode` or anything like this. Method is also just `add` which suggests nothing is done expect just adding a node to the tree ... somewhere. This is might be not clear for readers of this code so we need to improve this.

Second argument why I believe it would nice to move `add` to separate function is extensibility. What if we would like to use our tree with meta.jsons data struct for other algorithms? We can't because `add` has certain logic baked in.

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,159 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+func TestDeduplicateFilter_Filter(t *testing.T) {
+	for _, tcase := range []struct {
+		name     string
+		input    map[ulid.ULID]tsdb.BlockMetaCompaction
```
Maybe just `map[ulid.ULID][]ulid.ULID`?

bwplotka commented on pkg/block/fetcher.go:
```diff
@@ -410,3 +412,95 @@ func (f *LabelShardedMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, sync
+		if metas[id] != nil {
+			f.DuplicateIDs = append(f.DuplicateIDs, id)
+		}
+		// TODO(khyati): Increment synced metrics with proper state.
```
todo

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,117 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+		{
+			name: "compacted block with sources in bucket",
+			input: map[ulid.ULID][]ulid.ULID{
+				ULID(4): []ulid.ULID{ULID(1), ULID(2), ULID(3)},
```
Let's put other blocks here that we don't expect to be compacted

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,117 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+		{
+			name: "two compacted blocks with same sources",
+			input: map[ulid.ULID][]ulid.ULID{
+				ULID(3): []ulid.ULID{ULID(1), ULID(2)},
```
ditto, let's mix it

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,117 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+		{
+			name: "two compacted blocks with overlapping sources",
+			input: map[ulid.ULID][]ulid.ULID{
+				ULID(4): []ulid.ULID{ULID(1), ULID(2)},
```
ditto

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,117 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+			name: "two compacted blocks with overlapping sources",
+			input: map[ulid.ULID][]ulid.ULID{
+				ULID(4): []ulid.ULID{ULID(1), ULID(2)},
+				ULID(5): []ulid.ULID{ULID(1), ULID(2), ULID(3)},
```
Can we reorder ULIDs in Source to check if that will work?

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,117 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+		{
+			name: "3 compacted blocks of level 2 and one compacted block of level 3 in bucket",
+			input: map[ulid.ULID][]ulid.ULID{
+				ULID(10): []ulid.ULID{ULID(1), ULID(2), ULID(3)},
```
ditto, let's mix with some block that should stay

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,117 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+			input: map[ulid.ULID][]ulid.ULID{
+				ULID(1): []ulid.ULID{ULID(1)},
+				ULID(5): []ulid.ULID{ULID(1), ULID(2)},
+				ULID(6): []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
```
Let's reorder those a bit to make it more tricky. Yes we put it in map anyway, but sometimes map tends to be.. accidently ordered (:

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,117 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+	} {
+		f := NewDeduplicateFilter()
+		if ok := t.Run(tcase.name, func(t *testing.T) {
+			synced := prometheus.NewGaugeVec(prometheus.GaugeOpts{}, []string{"state"})
```
We can assert on this metric! 

See our talk: https://are-you-testing-your-observability.now.sh/

Click `right` until 26 slide and check it out (:

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,117 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+				ULID(5): []ulid.ULID{ULID(1), ULID(2)},
+				ULID(6): []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
+				ULID(7): []ulid.ULID{ULID(1), ULID(2), ULID(3)},
+				ULID(8): []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
```
Let's also swap `ULID(7)` to `ULID(8)` (: Make it more tricky!

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +393,117 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+				ULID(7): []ulid.ULID{ULID(1), ULID(2), ULID(3)},
+				ULID(8): []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
+			},
+			expected: []ulid.ULID{
```
Can we add one more case?

The case would be something like 

`10 containing  1,2,6,7`
`11 containing  1,2,6,8`
(: 
And maybe:
`10 containing  1,2,6,7`
`11 containing  1,2,3`

khyatisoneji commented:
Added the requested changes:
1. implemented todo to add sync metric for filtered duplicates
2. Reordered ulids in test
3. Mixed compacted and non compacted ulids in test
4. Added requested test additions
5. added test to test metric for filtered duplicates.

bwplotka commented on pkg/block/fetcher_test.go:
```diff
@@ -393,3 +395,178 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+}
+
+func compareSliceWithMapKeys(tb testing.TB, m map[ulid.ULID]*metadata.Meta, s []ulid.ULID) {
+	_, file, line, _ := runtime.Caller(1)
```
Wow this is epic!  Probably bit over-engineered, but... thanks!  (:
